source/subscription-manager/subscription_manager.c

```diff
@@ -0,0 +1,166 @@
+/*
+ * FreeRTOS V202011.00
+ * Copyright (C) 2020 Amazon.com, Inc. or its affiliates.  All Rights Reserved.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of
+ * this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+ * the Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *
+ * https://www.FreeRTOS.org
+ * https://aws.amazon.com/freertos
+ *
+ */
+
+/**
+ * @file subscription_manager.c
+ * @brief Functions for managing MQTT subscriptions.
+ */
+
+/* Subscription manager header include. */
+#include "subscription_manager.h"
+
+
+bool addSubscription( SubscriptionElement_t * pxSubscriptionList,
+                      const char * pcTopicFilterString,
+                      uint16_t usTopicFilterLength,
+                      IncomingPubCallback_t pxIncomingPublishCallback,
+                      void * pvIncomingPublishCallbackContext )
+{
+    int32_t lIndex = 0;
+    size_t xAvailableIndex = SUBSCRIPTION_MANAGER_MAX_SUBSCRIPTIONS;
+    bool xReturnStatus = false;
+
+    if( ( pxSubscriptionList == NULL ) ||
+        ( pcTopicFilterString == NULL ) ||
+        ( usTopicFilterLength == 0U ) ||
+        ( pxIncomingPublishCallback == NULL ) )
+    {
+        LogError( ( "Invalid parameter. pxSubscriptionList=%p, pcTopicFilterString=%p,"
```

Comment:
This isn't a library, just a demo, so this error checking isn't really necessary if an assert would do

Reply:
I think even if this is not a library, the parameter validations are good to have.